swarm/src/lib.rs

```diff
@@ -1142,7 +1140,7 @@ where
                         ));
                     self.pending_swarm_events
                         .push_back(SwarmEvent::NewExternalAddrCandidate { address: addr });
-                } else {
+                } else if self.listeners().next().is_some() {
```

Comment:
While we are at it, can we replace this `else` with an early `return` instead?

Reply:
Also, shouldn't this apply to the entire match arm, even before address translation? If we don't have a listener, any candidate is definitely not an external address.